packages/eslint-plugin/src/rules/no-unsafe-type-assertion.ts

```diff
@@ -0,0 +1,156 @@
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+    const compilerOptions = services.program.getCompilerOptions();
+
+    function getAnyTypeName(type: ts.Type): string {
+      return tsutils.isIntrinsicErrorType(type) ? 'error typed' : '`any`';
+    }
+
+    function isObjectLiteralType(type: ts.Type): boolean {
+      return (
+        tsutils.isObjectType(type) &&
+        tsutils.isObjectFlagSet(type, ts.ObjectFlags.ObjectLiteral)
+      );
+    }
+
+    function checkExpression(
+      node: TSESTree.TSAsExpression | TSESTree.TSTypeAssertion,
+    ): void {
+      const expressionType = getConstrainedTypeAtLocation(
+        services,
+        node.expression,
+      );
+      const assertedType = getConstrainedTypeAtLocation(
+        services,
+        node.typeAnnotation,
+      );
+
+      // consider unchanged type as safe
+      const typeIsUnchanged = isTypeUnchanged(
+        compilerOptions,
+        expressionType,
+        assertedType,
+      );
+
+      if (typeIsUnchanged) {
+        return;
+      }
+
+      // handle cases when casting unknown ==> any.
+      if (isTypeAnyType(assertedType) && isTypeUnknownType(expressionType)) {
+        context.report({
+          node,
+          messageId: 'unsafeToAnyTypeAssertion',
+          data: {
+            type: '`any`',
+          },
+        });
+
+        return;
+      }
+
+      // handle cases when casting of an any expression.
+      const expressionAny = isUnsafeAssignment(
+        expressionType,
+        assertedType,
+        checker,
+        node.expression,
+      );
+
+      if (expressionAny) {
+        context.report({
+          node,
+          messageId: 'unsafeOfAnyTypeAssertion',
+          data: {
+            type: getAnyTypeName(expressionAny.sender),
+          },
+        });
+
+        return;
+      }
+
+      // handle cases when casting to an any type.
+      const assertedAny = isUnsafeAssignment(
+        assertedType,
+        expressionType,
+        checker,
+        node.typeAnnotation,
+      );
+
+      if (assertedAny) {
+        context.report({
+          node,
+          messageId: 'unsafeToAnyTypeAssertion',
+          data: {
+            type: getAnyTypeName(assertedAny.sender),
+          },
+        });
+
+        return;
+      }
+
+      // fallback to checking assignability
+      const nodeWidenedType = isObjectLiteralType(expressionType)
+        ? checker.getWidenedType(expressionType)
+        : expressionType;
+
+      const isAssertionSafe = checker.isTypeAssignableTo(
+        nodeWidenedType,
+        assertedType,
+      );
+
+      if (!isAssertionSafe) {
+        context.report({
+          node,
+          messageId: 'unsafeTypeAssertion',
+          data: {
+            type: checker.typeToString(assertedType),
+          },
+        });
+      }
+    }
```

Comment:
Following from https://github.com/typescript-eslint/typescript-eslint/pull/10051#discussion_r1794398363, my proposal would be: 

```suggestion
      // fallback to checking assignability
      const isAssertionSafe = checker.isTypeAssignableTo(
        expressionType,
        assertedType,
      );

      if (!isAssertionSafe) {
        const nodeWidenedType = isObjectLiteralType(expressionType)
          ? checker.getWidenedType(expressionType)
          : expressionType;

        const isUnsafeDueToExcessPropertyChecks =
          nodeWidenedType !== expressionType &&
          checker.isTypeAssignableTo(nodeWidenedType, assertedType);

        context.report({
          node,
          messageId: isUnsafeDueToExcessPropertyChecks
            ? 'unsafeTypeAssertionExcessProperties'
            : 'unsafeTypeAssertion',
          data: {
            type: checker.typeToString(assertedType),
          },
        });
      }
```

WDYT?

Reply:
Some of my thoughts:

I think we should definitely have a dedicated message if we are to report excess properties as unsafe.

While excess properties can be the cause of unsafe code (see #5955 and https://github.com/typescript-eslint/typescript-eslint/issues/5677#issuecomment-1251772990 for examples), the only example I can think of where this is unsafe with this rule is:

```ts
type X = { foo: string };

function foo(x: X) {
  // ...
}

// calling a function
foo({ foo: '', bar: '' } as X);

// assigning to a variable
const b: X = { foo: '', bar: '' } as X;
```

But since even assigning the object literal to a variable suppresses the issue (and removes the need for a type assertion), I'm not sure how helpful it would be to have the rule warn here (even though this is an unsafe assertion that can lead to bugs).

I personally think this may be a better fit in a separate rule (potentially `no-unsafe-excess-properties`, though this is a different take than what's suggested in #5955) or be baked into the other `no-unsafe-*` rules, so invalid code like the linked examples is warned.

Let me know if you prefer this to be part of the current rule - I'll be happy to update the PR accordingly.

Reply:
Ok, no, this is definitely fair. Let's leave it the way you have it!